ocaml/xenopsd/xc/readln.ml

```diff
@@ -23,25 +23,26 @@ let read fd =
              (Bytes.to_string pending)
           )
   | n ->
-      let data = Bytes.sub buffer 0 n in
-      let inpt = try Hashtbl.find input fd with Not_found -> Bytes.empty in
-      Hashtbl.replace input fd (Bytes.cat inpt data) ;
-      let rec loop msgs =
-        let data = Hashtbl.find input fd in
-        (* never fails *)
-        match Bytes.index data '\n' with
-        | exception Not_found ->
-            Ok (List.rev msgs)
-        | index ->
+      let rec loop msgs data =
+        match Bytes.index_opt data '\n' with
+        | None ->
+            (List.rev msgs, data)
+        | Some index ->
             let remain =
               Bytes.sub data (index + 1) (Bytes.length data - index - 1)
             in
-            Hashtbl.replace input fd remain ;
-            (* reset input *)
-            loop (Bytes.sub_string data 0 index :: msgs)
-        (* store msg *)
+            loop
+              (Bytes.sub_string data 0 index :: msgs)
+              remain (* reset input *)
+      in
+      let data = Bytes.sub buffer 0 n in
+      let inpt =
+        Option.value (Hashtbl.find_opt input fd) ~default:Bytes.empty
       in
-      loop []
+      let inp_data = Bytes.cat inpt data in
+      let res, data = loop [] inp_data in
+      Hashtbl.replace input fd data ;
```

Comment:
This substitution doesn't look correct to me. From what I can see, the previous code did not write to the hashtable when the input didn't contain `\n`, now inp_data is written to it. The code shold make sure the hastable is only written for the last index found.

Reply:
Previously, the hashtable was written to before entering the function (L28, without checking for the newline). Now, if I don't find a newline, I just return the input to the function and write it to the hashtable. It should be the same, I think.

Reply:
You're right, the previous code was very poor.